src/main/java/com/netflix/control/clutch/ExperimentalControlLoop.java

```diff
@@ -79,32 +84,48 @@ public ExperimentalControlLoop(ClutchConfiguration config, IActuator actuator, D
 
         Observable<Event> rps = events.filter(event -> event.getMetric() == Clutch.Metric.RPS);
 
-        Integrator integrator = new Integrator(initialSize, config.minSize, config.maxSize, config.integralDecay);
+        Integrator deltaIntegrator = new Integrator(0, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY, config.integralDecay);
 
         size
                 .takeUntil(timer)
                 .doOnNext(currentScale::set)
-                .doOnNext(integrator::setSum)
                 .doOnNext(__ -> cooldownTimestamp.set(System.currentTimeMillis()))
                 .doOnNext(n -> log.info("Clutch received new scheduling update with {} workers.", n))
                 .subscribe();
 
         return rps.withLatestFrom(lag, drops, Tuple::of)
                 .doOnNext(triple -> log.debug("Clutch received RPS: {}, Lag: {} (d {}), Drops: {}", triple._1.value, triple._2.value, triple._2.value - lastLag.get(), triple._3.value))
-                .map(triple -> {
-                  double lagDerivative  = triple._2.value - lastLag.get();
-                  lastLag.set(triple._2.value);
-                    return triple._1.value // RPS
-                            + (lagDerivative) // Lag derivative is of Max lag, so already per-worker.
-                            + (triple._3.value); // Drops are average, and thus per-worker.
-                })
+                .map(triple -> this.rpsMetricComputer.apply(triple._1.value, triple._2.value, triple._3.value))
                 .lift(new ErrorComputer(config.setPoint, true, config.rope._1, config.rope._2))
                 .lift(new PIDController(config.kp, config.ki, config.kd, 1.0, new AtomicDouble(1.0), config.integralDecay))
-                .lift(integrator)
+                .lift(deltaIntegrator)
                 .filter(__ -> this.cooldownMillis == 0 || cooldownTimestamp.get() <= System.currentTimeMillis() - this.cooldownMillis)
+                .map(delta -> this.scaleComputer.apply(this.currentScale.get(), delta, (long) config.minSize, (long) config.maxSize))
                 .filter(scale -> this.currentScale.get() != Math.round(Math.ceil(scale)))
                 .lift(actuator)
                 .doOnNext(scale -> this.currentScale.set(Math.round(Math.ceil(scale))))
+                .doOnNext(__ -> deltaIntegrator.setSum(0))
                 .doOnNext(__ -> cooldownTimestamp.set(System.currentTimeMillis()));
     }
+
+    /* For testing to trigger actuator on next event */
+    protected void setCooldownMillis(long cooldownMillis) {
+        this.cooldownMillis = cooldownMillis;
+    }
+
+    public static class DefaultRpsMetricComputer implements IRpsMetricComputer {
+        private double lastLag = 0;
+
+        public Double apply(Double rps, Double lag, Double drop) {
+            double lagDerivative  = lag - lastLag;
```

Comment:
What if we convert lag to lag/rps instead. If lag is not growing then the derivative maybe 0.

Reply:
This default implementation just replicate existing behavior. The idea is allow different implementation of RpsMetricComputer also variants can be tested and compared.